README.md
# estimacion
Estimacion mediante técnica de MHE
Primero se utilizaba el estimador original para estimar CD a partir de 3 casos: cd cte, cd como función de #M y CD como interpolación de != #M. Luego se le incorporó la estimación de CL. Surge una cuestión con CD porque estima CD vs M y CD vs t y eso no es caractístico del proyectil ya que si varía el ángulo durante la trayectoría también variará el CD. Por lo que se debe estimar CD0 y CDd2. Para esto se modifica las ecuaciones del estimador.

changelog.md
v0.0.01 -> Version DRAG. Se genera tres versiones de drag: i) drag cte ii) drag variable en función de  C_D = 0.125 * tanh (5*(M -0.9))+0.425 iii) drag variable según tabla de interpolación de McCoy, bullet sierra .308


octave/test_sintetico.m
clear all;close all;
% Caso Sintético - Cd manufacturado
% Datos
% Cd0
M = [0, 0.8, 0.85, 0.9, 0.95, 1.0, 1.05, 1.1, 1.2, 1.4, 1.6, 1.8, 2.0, 2.2, 2.5];
Cd0_exp = [0.14, 0.14, 0.142, 0.16, 0.24, 0.43, 0.449, 0.447, 0.434, 0.41, 0.385, 0.365, 0.35, 0.339, 0.32];

% Cdd2
Md2 = [0, 0.95, 1, 1.05, 1.1, 1.2, 1.4, 1.6, 1.8, 2.0, 2.2, 2.5];
Cdd2_exp = [2.9, 2.9, 3.0, 3.1, 3.6, 6.5, 7.6, 7.3, 6.8, 6.1, 5.4, 4.4];

% Cd = Cd0 + Cdd2 * delta^2

% Rango de Mach
mach = [2.5:-0.05:0]; % xq lo hace decreciente?
% Tiempo
t = [1/length(mach):1/length(mach):1];
% Ang de ataque max 10 grad
amax = 10*pi/180;% amax = deg2rad(10) convert degree2radiands
% Variación ang ataque vs tiempo +-10 grad
delta = amax*sin(4*t*pi); % delta rad
delta_deg = rad2deg(delta); % delta degree
sin_deg = rad2deg(delta/amax);
%figure
%plot(t,delta)

% Calculo Cd + una perturbación en los coef
Cd = (interp1(M, Cd0_exp, mach)+rand(1)*5e-4) + (interp1(Md2, Cdd2_exp, mach)+rand(1)*5e-4).*delta.^2;

% Interp del Cd0_exp(M) en func del vector mach, en realdiad tendria que ser Cd0 de otra corrida
Cd0_exp_interp =  interp1(M, Cd0_exp, mach);

x = Cd - Cd0_exp_interp;
Cddelta = zeros(1,length(x));

% Calculo de Cddelta solo si delta > 0.5 grad
for i = 1:length(x)
  if(x(i) > 0.5*pi/180)
    Cddelta(i) = (Cd(i) - Cd0_exp_interp(i)) / delta(i)^2;
  endif
endfor

% Figures
figure();subplot(2,1,1);plot(t,delta,'b',t,delta/amax,'r');title('Perturbacion radianes');legend('delta','sin');xlabel('t');subplot(2,1,2);plot(t,sin_deg,'b',t,delta_deg,'r');title('Perturbacion degree');legend('delta','sin');xlabel('t');print -dpng Figures/Perturbaciones_angulares.png
figure();plot(mach,Cd,'o-');title('Cd sintetico');print -dpng Figures/Cd_sintetico.png
figure();plot(mach,Cd0_exp_interp,'o-');title('Cd0 interpolado');print -dpng Figures/Cd0_interpolado.png
figure();plot(Md2,Cdd2_exp,mach,Cddelta,'+');title('Aprox Cddelta');legend('Real','Estimacion');xlabel('Mach');ylabel('Cddd2');print -dpng Figures/Aprox_Cdd2.png


octave/test_sint_adaptado_RBD.m
clear all;close all;
% ex Caso Sintético - Estimacion Cd0 y Cdd2 adaptado a Resu_RBD
% todo armado para que guarde con nombres automaticamente
tStart = tic; %timer de multivariables
% Datos Cd McCoy
% Cd0
M = [0, 0.8, 0.85, 0.9, 0.95, 1.0, 1.05, 1.1, 1.2, 1.4, 1.6, 1.8, 2.0, 2.2, 2.5]';
Cd0_exp = [0.14, 0.14, 0.142, 0.16, 0.24, 0.43, 0.449, 0.447, 0.434, 0.41, 0.385, 0.365, 0.35, 0.339, 0.32]';

% Cdd2
Md2 = [0, 0.95, 1, 1.05, 1.1, 1.2, 1.4, 1.6, 1.8, 2.0, 2.2, 2.5]';
Cdd2_exp = [2.9, 2.9, 3.0, 3.1, 3.6, 6.5, 7.6, 7.3, 6.8, 6.1, 5.4, 4.4]';

% Cd = Cd0 + Cdd2 * delta^2

% Colocar nombre del caso a analziar el CD
caso = 'B06_alpha15';
% maquinaria para cargar datos en funcion del nombre del caso
folder_path=('/home/zeeburg/Documents/CIMEC/Tesis/estimacion/Resu_RBD/');
path_file = fullfile([folder_path],['Force_coef_proc_',caso,'.txt']);
fprintf('\nSe cargan los datos de CD del archivo:\n%s\n',path_file)

# Time,   Mach,     alfa,     beta,     delta2,     Cd,     CL_alfa,     Cn_p_alfa,     Cn_q_alfa 
data = load(path_file);
mach_cd = data(:,2);
alfa_cd = data(:,3);
delta2 = data(:,5);
Cd = data(:,6);

% Cd0 sale de corrida previa a delta 0, B05
data0 = load('/home/zeeburg/Documents/CIMEC/Tesis/estimacion/Resu_RBD/Force_coef_proc_B05.txt');
mach0 = data0(:,2);
Cd0 = data0(:,6);
% ver cual es el max de los min de los mach: mach o mach_cd
mach = [max(mach0):-0.05:min(mach0)];
%mach_cd = [max(mach_cd):-0.05:min(mach_cd)];

Cd0_interp =  interp1(mach0, Cd0, mach_cd);

x = Cd - Cd0_interp;
Cddelta = zeros(1,length(x));

% Calculo de Cddelta solo si delta > 0.5 grad
for i = 1:length(x)
  if(delta2(i) > 0.05*pi/180)
    Cddelta(i) = (Cd(i) - Cd0_interp(i)) / delta2(i);
  end
end

% Figures
figure();plot(mach_cd,Cd,'o-');
title_Cd = sprintf('Cd %s',caso);
title(title_Cd);xlabel('Mach','FontSize',14);ylabel('Cd','FontSize',14)
figname_Cd_interp = sprintf('Cd_%s',caso);xlim([0 2.5])
print(fullfile('Figures',[figname_Cd_interp,'.png']),'-dpng')
% gnumeric Cd
gnumeric_Cd_interp = [mach_cd Cd]; % se crea vector para exportar como csv
cvs_Cdinterp = sprintf('./csv/cvsgnum_%s',figname_Cd_interp);
csvwrite (cvs_Cdinterp,gnumeric_Cd_interp);


figure();plot(M,Cd0_exp,mach_cd,Cd0_interp,'o');
title_Cd0 = sprintf('Cd0 %s',caso);
title(title_Cd0);xlabel('Mach','FontSize',14);ylabel('Cd0','FontSize',14);legend('Real','Estimacion');
figname_Cd0_interp = sprintf('Aprox_Cd0_%s',caso);xlim([0 2.5])
print(fullfile('Figures',[figname_Cd0_interp,'.png']),'-dpng')
% gnumeric Cdd2
gnumeric_Cd0_interp = [mach_cd Cd0_interp]; % se crea vector para exportar como csv
cvs_Cd0interp = sprintf('./csv/cvsgnum_%s',figname_Cd0_interp);
csvwrite (cvs_Cd0interp,gnumeric_Cd0_interp);


figure();plot(Md2,Cdd2_exp,mach_cd,Cddelta,'+');
title_Cdd2=sprintf('Cdd2 %s',caso);
title(title_Cdd2);legend('Real','Estimacion');xlabel('Mach','FontSize',14);ylabel('Cdd2','FontSize',14);
figname_Cdd2 = sprintf('Aprox_Cdd2_%s',caso);xlim([0 2.5])
print(fullfile('Figures',[figname_Cdd2,'.png']),'-dpng')
% gnumeric Cdd2
gnumeric_Cdd2_interp = [mach_cd Cddelta']; % se crea vector para exportar como csv
cvs_Cdd2interp = sprintf('./csv/cvsgnum_%s',figname_Cdd2);
csvwrite (cvs_Cdd2interp,gnumeric_Cdd2_interp);


% gnumeric Cd0 McCoy
gnumeric_Cd0_McCoy = [M Cd0_exp]; % se crea vector para exportar como csv
%cvs_Cd0_McCoy = sprintf('cvsgnum_%s',figname_Cd0_interp);
csvwrite ('./csv/csvgnum_Cd0_McCoy',gnumeric_Cd0_McCoy);
% gnumeric Cdd2 McCoy
gnumeric_Cdd2_McCoy = [Md2 Cdd2_exp]; % se crea vector para exportar como csv
%cvs_Cdd2_McCoy = sprintf('cvsgnum_%s',figname_Cdd2);
csvwrite ('./csv/csvgnum_Cdd2_McCoy',gnumeric_Cdd2_McCoy);

time = toc(tStart);
fprintf('*-----------------------------------------------*\n')
fprintf('\n\nFIN! - OK - time = %d[s].\n',time)


estimacion_mhe.py
# -*- coding: utf-8 -*-
"""

@author: lgenzelis, ntrivisonno
"""

import numpy as np
import matplotlib.pyplot as plt
import mpctools
import casadi
from scipy import linalg

diam = 0.04
rho = 1.225
S = np.pi * (0.5 * diam) ** 2

CASO = 0

# carga dato una sola corrida
#
data = np.loadtxt('Resu_RBD/' + ['Forces_proc_B06.txt'][CASO], delimiter=',', skiprows=1)
#
# carga datos, solo drag
#data = np.loadtxt('Resu_RBD/' + ['Forces_proc_B03.txt', 'Forces_proc_B04.txt', 'Forces_proc_B06.txt'][CASO], delimiter=',', skiprows=1) #mail_nicolas/ hace alusion a la carpeta | skiprows=1 saltea la primer fila xq es el encabezado
#
# carga datos, drag + lift
#data = np.loadtxt('Resu_RBD/' + ['Forces_proc_C_C01.txt', 'Forces_proc_C_C02.txt', 'Forces_proc_C_C03.txt'][CASO], delimiter=',', skiprows=1) #mail_nicolas/ hace alusion a la carpeta | skiprows=1 saltea la primer fila xq es el encabezado
N = data.shape[0]

print(data.shape)

# Encabezado del txt:
# Time, alpha, beta, delta2, V_inf (= V_t), u(v_body_X), v(v_body_Y), w(v_body_Z), p, q, r, gx, gy, gz, FX_body, FY_body, FZ_body

alpha = data[:, 1]
beta = data[:, 2]
delta2 = data[:, 3]  # alpha2
vt = data[:, 4]
u = data[:, 5]  # vel_body_X
v = data[:, 6]  # vel_body_Y
w = data[:, 7]  # vel_body_Z
p = data[:, 8]
q = data[:, 9]
r = data[:, 10]
grav = data[:, 11:14]  # gx, gy, gz
F_body = data[:, 14:17]  # FX_body, FY_body, FZ_body

# F_body = F_body - grav  # FIXME

Ncoef = 2 # cant de coef a estimar
Ny = 3
Nw = Ncoef
Nv = Ny
Np = 3
Nt = 10  # horizonte
Nu = 0

Q = np.diag([1000.] * Ncoef)  # matrix de covarianza de ruido de proceso
R = np.diag([.1, .1, .1])     # matrix de covarianza de ruido de  medición
P = np.diag([10.] * Ncoef)    # matrix de covarianza de estimación inicial

Q_inv = linalg.inv(Q)
R_inv = linalg.inv(R)

def ode(x, u, w):
    dxdt = 0. + w
    return np.array(dxdt)

def meas(x, p):
    '''
    x[0]: Cd
    x[1]: Cl_alpha
    p[0]: vt
    p[1]: alpha
    p[2]: beta
    y[0]: Fx
    y[1]: Fy
    y[2]: Fz
    '''
    qdy = 0.5 * rho * p[0] ** 2
    y = casadi.SX.zeros(Ny)
    y[0] = -qdy * S * np.cos(p[1]) * np.cos(p[2]) * x[0]# + qdy * S * x[1] * (1-(np.cos(p[1])**2 * np.cos(p[2])**2))
    y[1] = -qdy * S * np.sin(p[2]) * x[0]# - qdy * S * x[1] * np.cos(p[1]) * np.cos(p[2]) * np.sin(p[2])
    y[2] = -qdy * S * np.sin(p[1]) * np.cos(p[2]) * x[0]# + qdy * S * x[1] * np.cos(p[1]) * np.sin(p[1]) * np.cos(p[2])**2
    assert Ny == 3
    return y

# no importa el Delta que ponga (por cómo es mi "ode")
F = mpctools.getCasadiFunc(ode, [Ncoef, Nu, Nw], ["x", "u", "w"], "F", rk4=True, Delta=.1)
H = mpctools.getCasadiFunc(meas, [Ncoef, Np], ["x", "p"], "H")

# Costo de etapa
def lfunc(w, v):
    return mpctools.mtimes(w.T, Q_inv, w) + mpctools.mtimes(v.T, R_inv, v)
l = mpctools.getCasadiFunc(lfunc, [Nw, Nv], ["w", "v"], "l")

# Costo de arrivo
def lxfunc(x, x0bar, Pinv):
    dx = x - x0bar
    if Ncoef == 1:
        return mpctools.mtimes(dx.T, dx) * Pinv
    else:
        return mpctools.mtimes(dx.T, Pinv, dx)
lx = mpctools.getCasadiFunc(lxfunc, [Ncoef, Ncoef, (Ncoef, Ncoef)], ["x", "x0bar", "Pinv"], "lx")

xhat = np.zeros((N, Ncoef))
yhat = np.zeros((N, Nv))
x0bar = np.zeros(Ncoef)
guess = {}


def ekf(f, h, x, u, w, y, p, P, Q, R, f_jacx=None, f_jacw=None, h_jacx=None):
    """
    EKF copiado del de mpctools, para adaptarlo a que H dependa de p también.

    Updates the prior distribution P^- using the Extended Kalman filter.

    f and h should be casadi functions. f must be discrete-time. P, Q, and R
    are the prior, state disturbance, and measurement noise covariances. Note
    that f must be f(x,u,w) and h must be h(x).

    If specified, f_jac and h_jac should be initialized jacobians. This saves
    some time if you're going to be calling this many times in a row, although
    it's really not noticable unless the models are very large. Note that they
    should return a single argument and can be created using
    mpctools.util.jacobianfunc.

    The value of x that should be fed is xhat(k | k-1), and the value of P
    should be P(k | k-1). xhat will be updated to xhat(k | k) and then advanced
    to xhat(k+1 | k), while P will be updated to P(k | k) and then advanced to
    P(k+1 | k). The return values are a list as follows

        [P(k+1 | k), xhat(k+1 | k), P(k | k), xhat(k | k)]

    Depending on your specific application, you will only be interested in
    some of these values.
    """

    # Check jacobians.
    if f_jacx is None:
        f_jacx = mpctools.util.jacobianfunc(f, 0)
    if f_jacw is None:
        f_jacw = mpctools.util.jacobianfunc(f, 2)
    if h_jacx is None:
        h_jacx = mpctools.util.jacobianfunc(h, 0)

    # Get linearization of measurement.
    C = np.array(h_jacx(x, p))
    yhat = np.array(h(x, p)).flatten()

    # Advance from x(k | k-1) to x(k | k).
    xhatm = x                                          # This is xhat(k | k-1)
    Pm = P                                             # This is P(k | k-1)
    L = linalg.solve(C.dot(Pm).dot(C.T) + R, C.dot(Pm)).T
    xhat = xhatm + L.dot(y - yhat)                     # This is xhat(k | k)
    P = (np.eye(Pm.shape[0]) - L.dot(C)).dot(Pm)       # This is P(k | k)

    # Now linearize the model at xhat.
    w = np.zeros(w.shape)
    A = np.array(f_jacx(xhat, u, w))
    G = np.array(f_jacw(xhat, u, w))

    # Advance.
    Pmp1 = A.dot(P).dot(A.T) + G.dot(Q).dot(G.T)       # This is P(k+1 | k)
    xhatmp1 = np.array(f(xhat, u, w)).flatten()     # This is xhat(k+1 | k)

    return [Pmp1, xhatmp1, P, xhat]

for k in range(N):
    N = {"x": Ncoef, "y": Ny, "p": Np, "u": Nu}
    N["t"] = min(k, Nt)
    tmin = max(0, k - Nt)
    tmax = k+1  # para que en los slice cuando ponga :tmax tome hasta k *inclusive*

    p_coefs = np.vstack((vt[tmin:tmax], alpha[tmin:tmax], beta[tmin:tmax])).T
    assert p_coefs.shape == (N["t"] + 1, Np)

    # Armo y llamo al solver. Si todavía no llené el horizonte, armo uno nuevo. Sino reúso el viejo.
    if k <= Nt:
        solver = mpctools.nmhe(f=F, h=H, u=np.zeros((tmax-tmin-1, Nu)), p=p_coefs,
                               y=F_body[tmin:tmax, :], l=l, N=N, lx=lx,
                               x0bar=x0bar, verbosity=0, guess=guess,
                               extrapar=dict(Pinv=linalg.inv(P)), inferargs=True)
    else:
        solver.par["Pinv"] = linalg.inv(P)
        solver.par["x0bar"] = x0bar
        solver.par["y"] = list(F_body[tmin:tmax, :])
        solver.par["p"] = list(p_coefs)
    sol = mpctools.callSolver(solver)
    print(("%3d: %s" % (k, sol["status"])))
    if sol["status"] != "Solve_Succeeded":
        break

    xhat[k] = sol["x"][-1]  # xhat( t  | t )
    yhat[k] = np.squeeze(H(xhat[k], p_coefs[-1]))

    # Armo el guess.
    guess = {}
    for var in set(["x","w","v"]).intersection(sol.keys()):
        guess[var] = sol[var].copy()
    # Actualizo el guess
    if k + 1 > Nt:
        for key in guess.keys():
            guess[key] = guess[key][1:, ...]  # Tiro el guess más viejo

        # EKF para actualizar la matriz de covarianza P
        [P, _, _, _] = ekf(F, H, x=sol["x"][0, :], u=np.zeros(Nu), w=sol["w"][0, ...], y=F_body[tmin], p=p_coefs[0], P=P, Q=Q, R=R)
        print("P: ", P)
        x0bar = sol["x"][1]

    # Repito el último guess como guess del nuevo instante de tiempo
    for key in guess.keys():
        guess[key] = np.concatenate((guess[key], guess[key][-1:]))

Cd_estim = xhat[:, 0]
Cl_estim = xhat[:, 1]

#
# carga datos 1 sola corrida
coefs_reales = np.loadtxt('Resu_RBD/' + ['Force_coef_proc_B06.txt'][CASO], delimiter=',', skiprows=1)

# Encabezado del txt:
# Time,   Mach,     alfa,     beta,     delta2,     Cd,     CL_alfa,     Cn_p_alfa,     Cn_q_alfa

t = coefs_reales[:, 0]
mach = coefs_reales[:, 1]
Cd_real = coefs_reales[:, 5]
Cl_real = coefs_reales[:, 6]


#
# Cd
#

# estimacion
# Grafico el coeficiente
f, ax = plt.subplots(2)
ax[0].plot(mach, Cd_real,'o', label='Cd Real')
ax[0].plot(mach, Cd_estim, label='Cd Estimado')
ax[0].legend()
ax[0].set_xlim([min(mach), max(mach)])
ax[0].set_title('Cd vs Mach')

ax[1].plot(t, Cd_real,'o', label='Cd Real')
ax[1].plot(t, Cd_estim, label='Cd Estimado')
ax[1].legend()
ax[1].set_xlim([0, max(t)])
ax[1].set_title('Cd vs tiempo')

plt.tight_layout()
plt.savefig('Figures/CD - ' + ['Caso 8', 'Caso 10', 'Caso 11'][CASO] + '.png', bbox_inches='tight')

## Grafico errores
f, ax = plt.subplots(2)
ax[0].plot(mach, Cd_real - Cd_estim)
ax[0].set_xlim([min(mach), max(mach)])
ax[0].set_title('Error en Cd vs Mach')

ax[1].plot(t, Cd_real - Cd_estim)
ax[1].set_xlim([0, max(t)])
ax[1].set_title('Error en Cd vs tiempo')

plt.tight_layout()
plt.savefig('Figures/CD_error - ' + ['Caso 8', 'Caso 10', 'Caso 11'][CASO] + '.png', bbox_inches='tight')
'''
#
# Cl
#
# Grafico el coeficiente
f, ax = plt.subplots(2)
ax[0].plot(mach, Cl_real, label='Cl Real')
ax[0].plot(mach, Cl_estim, label='Cl Estimado')
ax[0].legend()
ax[0].set_xlim([min(mach), max(mach)])
ax[0].set_title('Cl vs Mach')

ax[1].plot(t, Cl_real, label='Cl Real')
ax[1].plot(t, Cl_estim, label='Cl Estimado')
ax[1].legend()
ax[1].set_xlim([0, max(t)])
ax[1].set_title('Cl vs tiempo')

plt.tight_layout()
plt.savefig('Figures/Cl - ' + ['Caso 8', 'Caso 10', 'Caso 11'][CASO] + '.png', bbox_inches='tight')

## Grafico errores 
f, ax = plt.subplots(2)
ax[0].plot(mach, Cl_real - Cl_estim)
ax[0].set_xlim([min(mach), max(mach)])
ax[0].set_title('Error en Cl vs Mach')

ax[1].plot(t, Cl_real - Cl_estim)
ax[1].set_xlim([0, max(t)])
ax[1].set_title('Error en Cl vs tiempo')

plt.tight_layout()
plt.savefig('Figures/Cl_error - ' + ['Caso 8', 'Caso 10', 'Caso 11'][CASO] + '.png', bbox_inches='tight')

plt.show()
'''

plt.show()


octave/calc_cdd2.m
function [mach_cd,Cddelta] = calc_cdd2(data)
  % funcion que calcula Cddelta segun Resu de RBD
  % Cd = Cd0 + Cddelta * delta^2
  % data el el archivo de la corrida a leer el Cd para calcular el Cdd2
  % ej:data = load('/home/zeeburg/Documents/CIMEC/Tesis/estimacion/Resu_RBD/Force_coef_proc_B06_alpha80.txt');
	 
  mach_cd = data(:,2);
  alfa = data(:,3);
  alfa_deg = rad2deg(alfa);
  delta2 = data(:,5);
  Cd = data(:,6);
  
    % Interp del Cd0_exp(M) en func del vec mach
  % Cd0_exp_interp =  interp1(M, Cd0_exp, mach);

  % Cd0_exp_interp sale de corrida previa a delta 0, B05
  data0 = load('/home/zeeburg/Documents/CIMEC/Tesis/estimacion/Resu_RBD/Force_coef_proc_B05.txt');
  mach0 = data0(:,2);
  Cd0 = data0(:,6);
  % ver cual es el max de los min de los mach: mach o mach_cd
  mach = [max(mach0):-0.05:min(mach0)];
  %mach_cd = [max(mach_cd):-0.05:min(mach_cd)];
  
  %Cd_interp =  interp1(mach_cd, Cd, mach);
  Cd0_interp =  interp1(mach0, Cd0, mach_cd);

  x = Cd - Cd0_interp;
  Cddelta = zeros(1,length(x));
  
  % Calculo de Cddelta solo si delta > 0.5 grad
  for i = 1:length(x)
    if(x(i) > 0.05*pi/180)
      Cddelta(i) = abs((Cd(i) - Cd0_interp(i))) / delta2(i);
    end
  end
  
end


octave/main_plot_aoa_5_to_80.m
% simil test_sint_adaptado_CFD.m plotea Cdd2 from aoa = 5 to 80 degrees
% guarda a todas las variables cargadas para poder plotearlas juntas en una misma grafica
clear all;close all;clc
tStart = tic; % timer de multivariables
% Datos
% Cd0
M = [0, 0.8, 0.85, 0.9, 0.95, 1.0, 1.05, 1.1, 1.2, 1.4, 1.6, 1.8, 2.0, 2.2, 2.5];
Cd0_exp = [0.14, 0.14, 0.142, 0.16, 0.24, 0.43, 0.449, 0.447, 0.434, 0.41, 0.385, 0.365, 0.35, 0.339, 0.32];
% Cdd2
Md2 = [0, 0.95, 1, 1.05, 1.1, 1.2, 1.4, 1.6, 1.8, 2.0, 2.2, 2.5];
Cdd2_exp = [2.9, 2.9, 3.0, 3.1, 3.6, 6.5, 7.6, 7.3, 6.8, 6.1, 5.4, 4.4];

data = load('/home/zeeburg/Documents/CIMEC/Tesis/estimacion/Resu_RBD/Force_coef_proc_B06_alpha5.txt');
[mach_cd_alpha5,Cddelta_alpha5] = calc_cdd2(data);
gnumeric_Cdd2_alpha5 = [mach_cd_alpha5,Cddelta_alpha5']; 
csvwrite ('./cvs_5_to_80/cvs_Cdd2interp_alpha5',gnumeric_Cdd2_alpha5);

data = load('/home/zeeburg/Documents/CIMEC/Tesis/estimacion/Resu_RBD/Force_coef_proc_B06_alpha15.txt');
[mach_cd_alpha15,Cddelta_alpha15] = calc_cdd2(data);
gnumeric_Cdd2_alpha15 = [mach_cd_alpha15,Cddelta_alpha15']; 
csvwrite ('./cvs_5_to_80/cvs_Cdd2interp_alpha15',gnumeric_Cdd2_alpha15);

data = load('/home/zeeburg/Documents/CIMEC/Tesis/estimacion/Resu_RBD/Force_coef_proc_B06_alpha30.txt');
[mach_cd_alpha30,Cddelta_alpha30] = calc_cdd2(data);
gnumeric_Cdd2_alpha30 = [mach_cd_alpha30,Cddelta_alpha30']; 
csvwrite ('./cvs_5_to_80/cvs_Cdd2interp_alpha30',gnumeric_Cdd2_alpha30);

data = load('/home/zeeburg/Documents/CIMEC/Tesis/estimacion/Resu_RBD/Force_coef_proc_B06_alpha40.txt');
[mach_cd_alpha40,Cddelta_alpha40] = calc_cdd2(data);
gnumeric_Cdd2_alpha40 = [mach_cd_alpha40,Cddelta_alpha40']; 
csvwrite ('./cvs_5_to_80/cvs_Cdd2interp_alpha40',gnumeric_Cdd2_alpha40);

data = load('/home/zeeburg/Documents/CIMEC/Tesis/estimacion/Resu_RBD/Force_coef_proc_B06_alpha50.txt');
[mach_cd_alpha50,Cddelta_alpha50] = calc_cdd2(data);
gnumeric_Cdd2_alpha50 = [mach_cd_alpha50,Cddelta_alpha50']; 
csvwrite ('./cvs_5_to_80/cvs_Cdd2interp_alpha50',gnumeric_Cdd2_alpha50);

data = load('/home/zeeburg/Documents/CIMEC/Tesis/estimacion/Resu_RBD/Force_coef_proc_B06_alpha60.txt');
[mach_cd_alpha60,Cddelta_alpha60] = calc_cdd2(data);
gnumeric_Cdd2_alpha60 = [mach_cd_alpha60,Cddelta_alpha60']; 
csvwrite ('./cvs_5_to_80/cvs_Cdd2interp_alpha60',gnumeric_Cdd2_alpha60);

data = load('/home/zeeburg/Documents/CIMEC/Tesis/estimacion/Resu_RBD/Force_coef_proc_B06_alpha70.txt');
[mach_cd_alpha70,Cddelta_alpha70] = calc_cdd2(data);
gnumeric_Cdd2_alpha70 = [mach_cd_alpha70,Cddelta_alpha70']; 
csvwrite ('./cvs_5_to_80/cvs_Cdd2interp_alpha70',gnumeric_Cdd2_alpha70);

data = load('/home/zeeburg/Documents/CIMEC/Tesis/estimacion/Resu_RBD/Force_coef_proc_B06_alpha80.txt');
[mach_cd_alpha80,Cddelta_alpha80] = calc_cdd2(data);
gnumeric_Cdd2_alpha80 = [mach_cd_alpha80,Cddelta_alpha80']; 
csvwrite ('./cvs_5_to_80/cvs_Cdd2interp_alpha80',gnumeric_Cdd2_alpha80);


%figure();plot(Md2,Cdd2_exp,mach_cd_alpha80,Cddelta_alpha80,'+');title('Aprox Cddelta alfa=80');legend('Real','Estimacion');xlabel('Mach');ylabel('Cddd2');%print -dpng Figures/Aprox_Cdd2_alfa80.png
figure();plot(Md2,Cdd2_exp,'b');hold on;
plot(mach_cd_alpha5,Cddelta_alpha5,'or');
plot(mach_cd_alpha15,Cddelta_alpha15,'+');
plot(mach_cd_alpha30,Cddelta_alpha30,'*');
plot(mach_cd_alpha40,Cddelta_alpha40,'x');
plot(mach_cd_alpha50,Cddelta_alpha50,'s');
plot(mach_cd_alpha60,Cddelta_alpha60,'d');
plot(mach_cd_alpha70,Cddelta_alpha70,'^');
plot(mach_cd_alpha80,Cddelta_alpha80,'d')
title('Cdd2 - Cuadratic drag');xlabel('Mach');ylabel('Cdd2')
legend('Real','alpha=5','alpha=15','alpha=30','alpha=40','alpha=50','alpha=60','alpha=70','alpha=80','location','SouthEast');
print -dpng Aprox_Cdd2_alpha5to80.png
%legend('Real','alpha=15','alpha=30','alpha=40','alpha=50','alpha=60','alpha=70','alpha=80','location','SouthEast')%Outside')
%xlim([0 1.3]);legend('location','NorthWest')

time = toc(tStart);
fprintf('*-----------------------------------------------*\n')
fprintf('\n\nFIN! - OK - time = %d[s].\n',time)
